### DISCUSSION TITLE
UI text when hovering over a variable whose type is a callable Protocol

### DISCUSSION THREAD

**Discussion by aholmes** [microsoft/pylance-release#5356] [Enhancement]
### References

With reference to Python 3.10 as a minimum version.

When type-hinting callables, we run into limitations and need workarounds. For example, a type I will refer to has the signature `(config: Config, request_context_args: dict[Any, Any] | None = None) -> RequestContext`.

Please reference this code.

Using `Callable`:
```python
FlaskRequestConfigurable = Callable[[Config, dict[Any, Any] | None], RequestContext]
```

Using `Protocol`:
```python
class FlaskRequestConfigurable(Protocol):
   def __call__(
       self, config: Config, request_context_args: dict[Any, Any] | None = None
   ) -> RequestContext:
       ...
```

Using either type as a paramter:
```python
class test_foo(flask_request_configurable: FlaskRequestConfigurable):
    pass
```

Using either type as a variable:
```python
flask_request_configurable: FlaskRequestConfigurable
```

Real code examples can be found [here](https://github.com/uclahs-cds/BL_Python/blob/3b5deb5ba1dd8a9c653b1f5babc50c564e3cc5d5/src/web/test/unit/create_app.py#L70) and [here](https://github.com/uclahs-cds/BL_Python/blob/3b5deb5ba1dd8a9c653b1f5babc50c564e3cc5d5/src/web/test/unit/middleware/test_middleware.py#L87).

### The Problem

With `Callable`, we can _almost_ represent this as `Callable[[Config, dict[Any, Any] | None], RequestContext]`. We run into three issues:
- `Callable` does not allow parameter names
- `Callable` does not allow default parameter values
- At call sites, positional arguments only are expected and named-parameters raise type errors

However, one advantage with `Callable` is in how Pylance displays information about it when hovering over the type:

![image](https://github.com/microsoft/pylance-release/assets/277061/9ce298a7-cae4-49f0-8fa0-88f32b2daeca)

Common advice to work around the issues mentioned is to specify a `Protocol` class and the `__call__` method. Indeed this does fix the issues, however, the UI experience is degraded.

![image](https://github.com/microsoft/pylance-release/assets/277061/49399095-56d9-4a43-9a42-64a89f900f67)

And in fact the UI experience when hovering over a parameter (specifically) whose type is the callable is less than ideal in both scenarios.

![image](https://github.com/microsoft/pylance-release/assets/277061/56fe1980-5090-436b-bf0e-2361e29fb44a)

Interestingly, I just noticed that a _variable_ whose type is `Callable` is not displayed the same way as a _parameter_. This is not true with the `Protocol` workaround, which does display the same as when used as a parameter.

![image](https://github.com/microsoft/pylance-release/assets/277061/235bd140-5e08-41f9-b95b-bbf7d1dd8802)

### The Desired Enhancement

When a variable or parameter is a `Protocol` type with `__call__` defined, and when hovered over or otherwise displaying information about the variable or parameter, showing the full call signature would make its purpose and usage more obvious. Ideally this would look similar, if not identical, to when the variable or parameter is a function:

![image](https://github.com/microsoft/pylance-release/assets/277061/9066b356-47a7-4ede-a88d-7a7c6ac3f267)

**Comment by rchiodo** [comment]
Seems reasonable to me. Protocol derived Classes should be represented like a function in the UI, not a class. Or maybe at least show extra information considering they're protocols.
